### PR TITLE
Add savings gamification engine

### LINKED ISSUE #837: feat(api): savings streak and gamification engine with anti-cheat integrity
# feat(api): savings streak and gamification engine with anti-cheat integrity

## 📌 Overview

Nester has a savings streak concept — a `savings_streaks` table (migration 050), a streak notifier ([apps/api/internal/service/streak_notifier.go](apps/api/internal/service/streak_notifier.go)), and a domain model. Streaks and gamification are among the most effective behavioural levers in consumer savings: a visible streak, a level, a badge for hitting a milestone, all convert an abstract long-term goal into daily engagement. But gamification is only motivating if it is *fair* and *believable*. A streak that resets wrongly because of a timezone bug, or a badge a user can farm by depositing one stroop a hundred times, destroys the trust the mechanic depends on. The engine must compute streaks and rewards correctly across timezones and edge cases, and it must resist trivial gaming.

This issue builds a **gamification engine** that computes streaks, levels and achievements from real saving behaviour, with the correctness and anti-cheat rigour that makes the rewards mean something.

## 🎯 Objectives

- Compute savings streaks correctly across timezones, missed days and grace periods.
- Define levels and achievements from genuine saving behaviour, not gameable actions.
- Resist trivial gaming (dust deposits, deposit-withdraw churn) so rewards stay meaningful.
- Surface progress and next milestones so the mechanic actively motivates.
- Keep the engine event-driven and consistent with the rest of the platform's state.

## 🧩 Requirements

Streak computation is deceptively hard because "a day" depends on the user's timezone, and a naive server-UTC implementation resets streaks at the wrong midnight for most of the world. Compute streak days in the user's local timezone (stored on the profile), define what counts as a qualifying action for a day (a deposit of at least a minimum amount, or maintaining a balance — pick and document the rule), and support a grace period so a single missed day within a window does not nuke a long streak, which is both kinder and more retentive. Every streak transition must be reproducible from the event history, so a disputed streak can be explained rather than asserted.

Anti-cheat is essential. A qualifying action must require genuine saving: a minimum deposit size so dust does not count, and a rule that a deposit immediately followed by a withdrawal does not qualify (net saving, not gross deposit). Levels and achievements derive from durable metrics — total saved, longest streak, goals completed, tenure — not from repeatable actions. Any achievement that can be earned by a loop the user controls is a bug; enumerate each achievement's condition and confirm it cannot be farmed.

Levels and achievements: define a level curve from a durable score (e.g. a function of total-saved and consistency), and a set of achievements with clear, one-time conditions. Achievements are awarded idempotently — awarding the same achievement twice is impossible — mirroring the milestone bitmask pattern used for goals. Reaching a level or achievement can trigger a notification and, where designed, a reward through the goal-rewards or referral machinery, but keep the reward funding bounded and out of user principal.

Progress surfacing is what makes it motivating: expose current streak, streak status (safe/at-risk/broken), grace remaining, current level with progress to next, and the nearest unearned achievement with its condition. "You're one day from your longest-ever streak" is the product working.

The engine is event-driven: qualifying actions arrive as events (from the indexer/ledger when a deposit confirms), the engine updates streak and metric state, and evaluates level/achievement transitions. It must be idempotent under event replay (the indexer is at-least-once) so a replayed deposit does not double-count toward a streak or metric.

## ⚙️ Technical Notes

Build on the streak domain ([apps/api/internal/domain/savingsstreak/model.go](apps/api/internal/domain/savingsstreak/model.go)) and repository ([apps/api/internal/repository/postgres/savings_streak_repository.go](apps/api/internal/repository/postgres/savings_streak_repository.go)). Migrations in [apps/api/migrations/](apps/api/migrations/) add levels, achievements and per-user gamification state. Timezone comes from the user profile (migration 027). The engine consumes confirmed-deposit events from the indexer/ledger; idempotency uses the same dedup discipline as the indexer. Notifications route through the unified notification service (companion issue). Streak evaluation on day boundaries fits the scheduler (see [apps/api/internal/scheduler/goal_deadline_reminder.go](apps/api/internal/scheduler/goal_deadline_reminder.go) for a per-user scheduled sweep pattern), but prefer event-driven updates with a scheduled sweep only for at-risk/break detection.

Handler exposure for progress; frontend surfaces would be a small follow-up within the frontend budget. Reference the existing streak notifier to avoid duplicating its logic.

Tests: timezone correctness (a user near a date line does not lose a streak wrongly); grace-period behaviour; dust and churn not qualifying; idempotent replay not double-counting; achievement awarded exactly once. Follow existing service test patterns.

## 📦 Expected Outcome

Streaks compute correctly in each user's local time with a forgiving grace period, and every transition is reproducible from event history. Levels and achievements reflect genuine, durable saving behaviour and cannot be farmed by dust or churn. Achievements award exactly once. Users see their streak status, level progress and nearest milestone, turning the mechanic into active motivation. The engine is event-driven and safe under the indexer's at-least-once replay.

## ✅ Acceptance Criteria

- [ ] Streaks are computed in the user's local timezone with a documented qualifying-action rule and a grace period.
- [ ] Every streak transition is reproducible from event history.
- [ ] Qualifying actions require genuine net saving; a dust test and a deposit-then-withdraw test both fail to qualify.
- [ ] Levels derive from a durable score and achievements have one-time, non-farmable conditions, each enumerated and verified.
- [ ] Achievements are awarded idempotently; a replay test proves no double-award and no double-count toward streaks/metrics.
- [ ] Progress (streak status, grace remaining, level progress, nearest achievement) is exposed via the API.
- [ ] Level/achievement events route through the unified notification service.
- [ ] Tests cover timezone edge cases, grace periods, anti-cheat and idempotent replay.

## 🚨 Contribution Rule

- All PRs MUST target the `dev` branch
- PRs opened against `main` will be automatically rejected
- Ensure your branch is created from `dev`

### PR DESCRIPTION
**Pull request by kitWarse**
Summary
- Adds a savings gamification engine for confirmed net savings activity, including timezone-aware daily streaks, grace handling, durable score/levels, one-time achievements, and anti-cheat qualification rules.

Issue
- Closes #837

Changes
- Added domain engine, documented qualification rules, and focused tests for local-day streaks, grace usage, dust/churn rejection, replay behavior, and achievement progress.
- Added Postgres persistence for gamification state, processed event dedupe, and one-time achievement awards.
- Added a service layer that records confirmed deposit events idempotently, emits achievement/level notifications, and exposes progress reads.
- Added authenticated progress API route at `GET /api/v1/users/savings-gamification/progress`.
- Wired split savings deposits into the gamification recorder and registered the new service/handler in the API bootstrap.
- Added migrations for user timezone, gamification state, event log, and achievement tables.

Validation
- Ran `git diff --check` successfully.
- Not run: `gofmt`, `go test`, or broader API validation because local `go`/`gofmt` are not installed on PATH in this environment.

Notes
- Per request, I did not spend time solving CI/environment setup.

## Summary by CodeRabbit

* **New Features**
  * Added savings-streak gamification with timezone-aware streak days, grace periods, durable scoring, levels, and one-time achievement unlocking.
  * Confirmed deposits now drive streak progression and related milestones.
  * Added a new endpoint to view savings gamification progress: `GET /api/v1/users/savings-gamification/progress`.
  * Achievement and level-up notifications now support replay/idempotent handling.
* **Documentation**
  * Updated the savings-streak gamification rules (qualification, streak/grace logic, scoring, and one-time awarding).
* **Tests**
  * Added unit tests for timezone handling, grace behavior, qualification rejection, and replay protection.

### CODE REVIEW

File: apps/api/internal/repository/postgres/savings_gamification_repository.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+			grace_used_for_day = EXCLUDED.grace_used_for_day,
+			total_saved = EXCLUDED.total_saved,
+			goals_completed = EXCLUDED.goals_completed,
+			current_level = EXCLUDED.current_level,
+			durable_score = EXCLUDED.durable_score,
+			awarded_achievements = EXCLUDED.awarded_achievements,
+			updated_at = NOW()
+	`, state.UserID, state.CurrentStreakDays, state.LongestStreakDays, state.LastQualifiedDay, state.GraceUsedForDay, state.TotalSaved.String(), state.GoalsCompleted, state.CurrentLevel, state.DurableScore.String(), pq.StringArray(state.AwardedAchievements))
+	return err
+}
```
Comment (Depo-dev) [inline apps/api/internal/repository/postgres/savings_gamification_repository.go:126]: This is a read-modify-write on `savings_gamification_state` with no transaction or row lock, so two deposits landing for the same user at the same time will lose one of the updates. Streak counts and scores are exactly the kind of thing users notice going backwards.

Same shape of problem in `savings_gamification_service.go` around line 104, where the transition isn't atomic either. Take the row under `SELECT ... FOR UPDATE` and do the read, the decision and the write in one transaction.

While you're in here: line 102 swallows the error from `RowsAffected()` and treats it as success, so a failed dedupe insert reads as a successful one. That's the guard #837 relies on for idempotent replay.

File: apps/api/internal/domain/savingsstreak/gamification.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+	AchievementWeekStreak   = "week_streak"
+	AchievementMonthStreak  = "month_streak"
+	AchievementHundredSaved = "hundred_saved"
+	AchievementGoalFinisher = "goal_finisher"
+)
+
+var (
+	ErrInvalidTimezone = errors.New("invalid user timezone")
+	ErrDuplicateEvent  = errors.New("duplicate gamification event")
+)
```
Comment (Depo-dev) [inline apps/api/internal/domain/savingsstreak/gamification.go:27]: Same-day and out-of-order events aren't handled correctly. Two deposits on the same local day can double-count toward the streak, and an event arriving after a later one has already been processed can move the streak backwards or re-award an achievement.

#837 asks that every streak transition be reproducible from event history and that a replay prove no double-award and no double-count. Neither holds while ordering is assumed. Sort or reject by event timestamp rather than arrival order, and make same-day accumulation explicit.

Smaller one at line 320: `GraceDays` is read as a boolean flag rather than a count, so configuring two grace days behaves the same as one.

File: apps/api/internal/service/savings_goal_service.go [hunk 3]
```diff
@@ -911,6 +921,7 @@ func (s *SavingsGoalService) DepositSplit(ctx context.Context, userID uuid.UUID,
 			ProgressPct:   pct,
 		}
 	}
+	s.recordGamificationDeposit(ctx, userID, in.TotalAmount, deposits, completedGoals(goals, results))
```
Comment (Depo-dev) [inline apps/api/internal/service/savings_goal_service.go:1047]: This computes all completed goals rather than newly completed ones, so every deposit re-fires the goal-finisher achievement path for goals that were already finished. The one-time award guard downstream saves you from a duplicate award, but you're doing the work and emitting the signal every time.

At line 950, gamification persistence failures are discarded silently. If recording the event fails, the deposit still succeeds and the streak quietly diverges from reality â€” which is the one thing #837 says must stay reproducible. At minimum log it; better, make it retryable through the job queue.

File: apps/api/internal/domain/savingsstreak/gamification_test.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	replayedState, replay, err := engine.Apply(state, event)
+	if err != nil {
+		t.Fatalf("Apply replay error = %v", err)
+	}
+	if len(replay.AwardedAchievements) != 0 || !replayedState.TotalSaved.Equal(state.TotalSaved) {
+		t.Fatalf("replay awards/state = %v/%s, want no awards/no double count", replay.AwardedAchievements, replayedState.TotalSaved)
+	}
+}
```
Comment (Depo-dev) [inline apps/api/internal/domain/savingsstreak/gamification_test.go:122]: The tests cover the right ideas â€” timezone-local day boundaries, grace usage, dust and churn rejection, replay. That's genuinely the correct list, and the dust plus deposit-then-withdraw cases are the ones #837 specifically asks for.

Two gaps: no coverage for multiple deposits in the same local day, and none for out-of-order delivery. Those are the two cases the concurrency issues above actually show up in, so please add them.

Also, the PR description says `gofmt` and `go test` were never run because Go isn't on PATH. Please get that working before the next round â€” I'd rather review tests that have executed.

File: apps/api/migrations/060_create_savings_gamification.up.sql [hunk 5]
```diff
@@ -0,0 +1,40 @@
+ALTER TABLE users ADD COLUMN IF NOT EXISTS timezone TEXT NOT NULL DEFAULT 'UTC';
+
+CREATE TABLE IF NOT EXISTS savings_gamification_state (
+    user_id UUID PRIMARY KEY REFERENCES users(id) ON DELETE CASCADE,
+    current_streak_days INTEGER NOT NULL DEFAULT 0,
+    longest_streak_days INTEGER NOT NULL DEFAULT 0,
+    last_qualified_day TEXT NOT NULL DEFAULT '',
+    grace_used_for_day TEXT NOT NULL DEFAULT '',
```
Comment (Depo-dev) [inline apps/api/migrations/060_create_savings_gamification.up.sql:8]: The `API (Go)` failure isn't yours. `dev` has two migrations at prefix 059 and the collision guard trips on every Go PR â€” I'm fixing that on `dev` now, so rebase once it lands and re-check that 060 is still unclaimed, since several open PRs want it.

Minor schema note while you're here: dates stored as `TEXT` with an empty-string sentinel would be better as nullable `DATE`. It's the kind of thing that's cheap now and annoying to migrate later.

File: apps/api/cmd/api/main.go [hunk 6]
```diff
@@ -487,6 +487,14 @@ func run() error {
 	savingsStreakRepo := postgres.NewSavingsStreakRepository(db)
 	savingsGoalSvc.SetStreakRepository(savingsStreakRepo)
 	savingsGoalSvc.SetStreakNotifier(service.DispatcherStreakMilestoneNotifier{Dispatcher: notificationDispatcher2})
+	savingsGamificationRepo := postgres.NewSavingsGamificationRepository(db)
+	savingsGamificationSvc := service.NewSavingsGamificationService(
+		savingsGamificationRepo,
+		service.DispatcherGamificationNotifier{Dispatcher: notificationDispatcher2},
+	)
```
Comment (Depo-dev) [inline apps/api/cmd/api/main.go:663]: Worth saying: this is the best-integrated of the three engine PRs on the board right now. The service is constructed in the bootstrap, the route is registered, and split savings deposits actually call into the recorder â€” so this is a working feature rather than a package waiting for someone to wire it up. That counts for a lot.

The concurrency and ordering issues above are what's holding it back, and they're all fixable. CodeRabbit has a few more small ones across the same files; please pick up the ones you agree with while you're in there.

**Review by Depo-dev — CHANGES_REQUESTED** [review]

**Review by 0xDeon — CHANGES_REQUESTED** [review]
This is conflicting with dev at the moment. Please rebase and resolve, then re-request review, streak/gamification logic is fiddly enough that I'd rather review it against a clean base.
